pkg/eventcollector/eventcollector.go

```diff
@@ -288,6 +308,67 @@ func (c *EventCollector) retrieveEvents() error {
 	return err
 }
 
+// CheckRetrievedEventsForNewsroom checks for new newsrooms from filtered events
+func (c *EventCollector) CheckRetrievedEventsForNewsroom(pastEvents []*model.Event) error {
+	log.Infof("Checking for new newsrooms in filterer")
+	existingFiltererNewsroomAddr := c.getExistingNewsroomFilterers()
+	existingWatcherNewsroomAddr := c.getExistingNewsroomWatchers()
+	watchersToAdd := map[common.Address]*watcher.NewsroomContractWatchers{}
+	for _, event := range pastEvents {
+		if event.EventType() == "ApplicationWhitelisted" {
+			newsroomAddr, ok := event.EventPayload()["ListingAddress"].(common.Address)
+			if !ok {
+				return fmt.Errorf("Cannot get newsroomAddr from eventpayload")
+			}
+			if _, ok := existingFiltererNewsroomAddr[newsroomAddr]; !ok {
+				newFilterer := filterer.NewNewsroomContractFilterers(newsroomAddr)
+				c.filterTriggers = append(c.filterTriggers, newFilterer)
+			}
+			if _, ok := existingWatcherNewsroomAddr[newsroomAddr]; !ok {
+				newWatcher := watcher.NewNewsroomContractWatchers(newsroomAddr)
+				watchersToAdd[newsroomAddr] = newWatcher
+			}
+		}
+		if event.EventType() == "ListingRemoved" {
+			newsroomAddr, ok := event.EventPayload()["ListingAddress"].(common.Address)
+			if !ok {
+				return fmt.Errorf("Cannot get newsroomAddr from eventpayload")
+			}
+			watchersToAdd[newsroomAddr] = nil
+		}
+	}
+	for addr, watcher := range watchersToAdd {
+		if watcher != nil {
+			log.Infof("Adding Newsroom watcher for %v", addr.Hex())
+			c.watchers = append(c.watchers, watcher)
+		} else {
+			log.Infof("Not adding %v to watchers because it was removed.", addr.Hex())
```

Comment:
Will there be a case we need to see if the watcher is in the `c.watchers` list to remove it?

Reply:
Does it go through `c.watchers` and remove the watcher from the list?

Reply:
no, I don't add it to c.watchers until I make sure that it's ok